components/camera/videosource/webcam.go

```diff
@@ -193,55 +199,232 @@ func makeConstraints(attrs *WebcamAttrs, debug bool, logger golog.Logger) mediad
+				continue
+			case <-timeoutSignal:
+				logger.Debugw("unhealthy listener; restarting", "camera.connected", isConnected)
+				startWard()
+				continue
+			}
+		}
+	}
+
+	label := attrs.Path
+	if label == "" {
+		label, err = getLabelFromCamera(cam)
+		if err != nil {
+			logger.Debugw("cannot find label for camera", "error", err)
+		}
+	}
+
+	// waits for the camera to disconnect then closes it.
+	disconnectWard := func(ctx context.Context, pulseInterval time.Duration) <-chan status {
+		heartbeat := make(chan status)
+		go func() {
+			logger.Debugf("starting camera disconnect ward with pulse interval %v", pulseInterval)
 			defer func() {
-				if err := cam.Close(ctx); err != nil {
-					logger.Debugw("failed to close camera", "error", err)
+				if err := recover(); err != nil {
+					debug.PrintStack()
+					golog.Global().Errorw("panic while running disconnect ward", "error", err)
 				}
+				logger.Debugf("closing camera %v", label)
+				goutils.UncheckedError(cam.Close(ctx))
+				close(heartbeat)
 			}()
+
 			for {
-				if goutils.SelectContextOrWait(ctx, 500*time.Millisecond) {
-					// mediadevices connects to the OS to get the properties for a driver. If the OS no longer knows
-					// about a specific driver then properties will be empty, and we can safely assume the driver no
-					// longer exists and should be closed.
-					if len(gostream.PropertiesFromMediaSource[image.Image, prop.Video](src)) == 0 {
-						logger.Debug("closing camera")
-						return
+				if !goutils.SelectContextOrWait(ctx, pulseInterval) {
+					return
+				}
+
+				select {
+				case <-ctx.Done():
+					return
+				case heartbeat <- monitoring:
+				default:
+				}
+
+				// If we cannot get the source or driver from our camera then we cannot know if it's connected.
+				// It's better to not assume it is or isn't and retry.
+				src, err := camera.SourceFromCamera(cam)
+				if err != nil {
+					logger.Debugw("cannot get source from camera", "error", err)
+					continue
+				}
+
+				d, err := gostream.DriverFromMediaSource[image.Image, prop.Video](src)
+				if err != nil {
+					logger.Debugw("cannot get media source from driver", "error", err)
+					continue
+				}
+
+				// https://github.com/pion/mediadevices connects to the OS to get the properties for a driver. If the
+				// OS no longer knows about a specific driver then properties will be empty, and we can safely assume the
+				// driver no longer exists and should be closed.
+				if len(d.Properties()) == 0 {
+					select {
+					case <-ctx.Done():
+					case heartbeat <- done:
 					}
-				} else {
 					return
 				}
 			}
-		})
-
-		return cam, err
+		}()
+		return heartbeat
 	}
 
-	source, err := gostream.GetAnyVideoSource(constraints, logger)
-	if err != nil {
-		return nil, errors.Wrapf(err, "found no webcams")
+	// waits for the camera to reconnect then repairs any references to it
+	reconnectWard := func(ctx context.Context, pulseInterval time.Duration) <-chan status {
```

Comment:
Echoing Eric, it does feel like a ManagedGo with PanicCapturingGoWithCallBack might work here too -- where the callback function can try to reconnect the camera, or do something else, depending on the error/panic that was recovered. The goroutine in the ManagedGo is essentially the steward that launches and keeps tracks of panics/errors from the ward.

Reply:
Done!